# 1. Import Libraries & Dataframes

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Set path
path = r'C:\Users\Ryan Pares\Instacart Basket Analysis'

In [3]:
# Import dataframes ()
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 data' , 'prepared data' , 'orders_products_merged_variables.pkl'))

In [6]:
# Check shape of data
df = ords_prods_merge
df.shape

(32404859, 21)

In [7]:
df.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  ...  \
0                     NaN       196.0                1.0        0.0  ...   
1                    15.0       196.0                1.0        1.0  ...   
2                    21.0       196.0                1.0        1.0  ...   
3                    29.0       196.0                1.0        1.0  ...   
4                    28.0       196.0                1.0        1.0  ...   

  product_name aisle_id  department_id  prices  match    price_range_loc  \
0         Soda       77              7     9.0   both  Mid-range product   
1         Soda       77              7     9.0   both  Mid-range product   
2         Soda       77              7     9.0   both  Mid-range product   
3         Soda       77              7     9.0   both  Mid-range product   
4         Soda       77              7     9.0   both  Mid-range product   

      busiest_day     Busiest_day    busiest_days busiest_period_of_day  
0  Regularly busy  Regularly busy  Regularly busy        Average orders  
1  Regularly busy  Regularly busy    Slowest days        Average orders  
2  Regularly busy  Regularly busy    Slowest days           Most orders  
3      Least busy      Least busy    Slowest days        Average orders  
4      Least busy      Least busy    Slowest days           Most orders  

[5 rows x 21 columns]

# 2. Find the aggregated mean of the “order_number” column grouped by “department_id”

In [ ]:
# Grouping by 'department_id' and 'order_number', and Calculating Mean using agg() Function
df.groupby('department_id').agg({'order_number' : ['mean']})

# 3. Results for the entire dataframe v/s the subset
#### The full dataset demonstrates a more regular ordering trend throughout the departments when contrasted with the subset. A striking contrast is seen in Department 21, where the mean is notably higher in the subset (25.53) compared to the full dataset (22.90). Similarly, Department 10 shows a significant uptick in the complete dataframe (20.19) against the subset (18.68).  When examining the subset against the entire dataframe, it's observed that the latter displays a greater average of order numbers across departments, reflecting a more comprehensive capture of customer activities.

# 4. Create a loyalty flag for existing customers using the transform() and loc() functions.

In [8]:
# Aggregating Data with transform() Function
df['max_order'] = df.groupby(['user_id'])['order_number'].transform(np.max)

In [9]:
# Checking the Output
df.head(15)

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0    2539329        1    prior             1          2                  8   
1    2398795        1    prior             2          3                  7   
2     473747        1    prior             3          3                 12   
3    2254736        1    prior             4          4                  7   
4     431534        1    prior             5          4                 15   
5    3367565        1    prior             6          2                  7   
6     550135        1    prior             7          1                  9   
7    3108588        1    prior             8          1                 14   
8    2295261        1    prior             9          1                 16   
9    2550362        1    prior            10          4                  8   
10   2968173       15    prior            15          1                  9   
11   1870022       15    prior            17          2                 16   
12   1911383       15    prior            18          2                 11   
13   2715276       15    prior            21          1                  9   
14    487368       15    prior            22          1                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered  ...  \
0                      NaN       196.0                1.0        0.0  ...   
1                     15.0       196.0                1.0        1.0  ...   
2                     21.0       196.0                1.0        1.0  ...   
3                     29.0       196.0                1.0        1.0  ...   
4                     28.0       196.0                1.0        1.0  ...   
5                     19.0       196.0                1.0        1.0  ...   
6                     20.0       196.0                1.0        1.0  ...   
7                     14.0       196.0                2.0        1.0  ...   
8                      0.0       196.0                4.0        1.0  ...   
9                     30.0       196.0                1.0        1.0  ...   
10                     7.0       196.0                2.0        0.0  ...   
11                     8.0       196.0                6.0        1.0  ...   
12                     7.0       196.0                1.0        1.0  ...   
13                     7.0       196.0                1.0        1.0  ...   
14                    14.0       196.0                1.0        1.0  ...   

   aisle_id department_id  prices  match    price_range_loc     busiest_day  \
0        77             7     9.0   both  Mid-range product  Regularly busy   
1        77             7     9.0   both  Mid-range product  Regularly busy   
2        77             7     9.0   both  Mid-range product  Regularly busy   
3        77             7     9.0   both  Mid-range product      Least busy   
4        77             7     9.0   both  Mid-range product      Least busy   
5        77             7     9.0   both  Mid-range product  Regularly busy   
6        77             7     9.0   both  Mid-range product  Regularly busy   
7        77             7     9.0   both  Mid-range product  Regularly busy   
8        77             7     9.0   both  Mid-range product  Regularly busy   
9        77             7     9.0   both  Mid-range product      Least busy   
10       77             7     9.0   both  Mid-range product  Regularly busy   
11       77             7     9.0   both  Mid-range product  Regularly busy   
12       77             7     9.0   both  Mid-range product  Regularly busy   
13       77             7     9.0   both  Mid-range product  Regularly busy   
14       77             7     9.0   both  Mid-range product  Regularly busy   

       Busiest_day    busiest_days busiest_period_of_day max_order  
0   Regularly busy  Regularly busy        Average orders        10  
1   Regularly busy    Slowest days        Average orders        10  
2   Regularly busy    Slowest days           Most 

In [10]:
# Creating a 'Loyalty' Label Flag using loc Function
df.loc[df['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
df.loc[(df['max_order'] <= 40) & (df['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
df.loc[df['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [11]:
# Examine Value Counts on 'loyalty_flag'
df['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [12]:
# Checking Updated Output
df.head(15)

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0    2539329        1    prior             1          2                  8   
1    2398795        1    prior             2          3                  7   
2     473747        1    prior             3          3                 12   
3    2254736        1    prior             4          4                  7   
4     431534        1    prior             5          4                 15   
5    3367565        1    prior             6          2                  7   
6     550135        1    prior             7          1                  9   
7    3108588        1    prior             8          1                 14   
8    2295261        1    prior             9          1                 16   
9    2550362        1    prior            10          4                  8   
10   2968173       15    prior            15          1                  9   
11   1870022       15    prior            17          2                 16   
12   1911383       15    prior            18          2                 11   
13   2715276       15    prior            21          1                  9   
14    487368       15    prior            22          1                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered  ...  \
0                      NaN       196.0                1.0        0.0  ...   
1                     15.0       196.0                1.0        1.0  ...   
2                     21.0       196.0                1.0        1.0  ...   
3                     29.0       196.0                1.0        1.0  ...   
4                     28.0       196.0                1.0        1.0  ...   
5                     19.0       196.0                1.0        1.0  ...   
6                     20.0       196.0                1.0        1.0  ...   
7                     14.0       196.0                2.0        1.0  ...   
8                      0.0       196.0                4.0        1.0  ...   
9                     30.0       196.0                1.0        1.0  ...   
10                     7.0       196.0                2.0        0.0  ...   
11                     8.0       196.0                6.0        1.0  ...   
12                     7.0       196.0                1.0        1.0  ...   
13                     7.0       196.0                1.0        1.0  ...   
14                    14.0       196.0                1.0        1.0  ...   

   department_id prices  match    price_range_loc     busiest_day  \
0              7    9.0   both  Mid-range product  Regularly busy   
1              7    9.0   both  Mid-range product  Regularly busy   
2              7    9.0   both  Mid-range product  Regularly busy   
3              7    9.0   both  Mid-range product      Least busy   
4              7    9.0   both  Mid-range product      Least busy   
5              7    9.0   both  Mid-range product  Regularly busy   
6              7    9.0   both  Mid-range product  Regularly busy   
7              7    9.0   both  Mid-range product  Regularly busy   
8              7    9.0   both  Mid-range product  Regularly busy   
9              7    9.0   both  Mid-range product      Least busy   
10             7    9.0   both  Mid-range product  Regularly busy   
11             7    9.0   both  Mid-range product  Regularly busy   
12             7    9.0   both  Mid-range product  Regularly busy   
13             7    9.0   both  Mid-range product  Regularly busy   
14             7    9.0   both  Mid-range product  Regularly busy   

       Busiest_day    busiest_days busiest_period_of_day max_order  \
0   Regularly busy  Regularly busy        Average orders        10   
1   Regularly busy    Slowest days        Average orders        10   
2   Regularly busy    Slowest days           Most orders        10   
3       Least busy    Slowest days        Average orders        10   
4       Least busy    Slowest days           Most orders        10 

# 5. Check if there’s a difference between the spending habits of the 3 types of customers (Loyal Customer, Regular Customer, and New Customer)

In [13]:
# Determining Basic Statistics of Spending Habits by Loyalty Category
df.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

#### Loyal customers generally opt for more affordable purchases, averaging 10.39 per item. In contrast, new customers appear to spend more freely, with their average expenditure per item being 13.29. Regular customers' spending habits are intermediate, with an average spend of 12.50 per item. While the minimum spend across all customer types is 1.0, the presence of a maximum spend of 99,999 could indicate the presence of outliers or inaccuracies within the data.

# 6. Create a spending flag for each user based on the average price across all their orders

In [14]:
# Aggregating Data with transform() Function
df['avg_product_price'] = df.groupby(['user_id'])['prices'].transform(np.mean)

In [15]:
# Checking Relevant Columns 
df[['user_id', 'avg_product_price']].head(15)

user_id  avg_product_price
0         1           6.367797
1         1           6.367797
2         1           6.367797
3         1           6.367797
4         1           6.367797
5         1           6.367797
6         1           6.367797
7         1           6.367797
8         1           6.367797
9         1           6.367797
10       15           3.980556
11       15           3.980556
12       15           3.980556
13       15           3.980556
14       15           3.980556

In [16]:
# Creating a 'Spending' Label Flag using loc Function
df.loc[df['avg_product_price'] >= 10, 'spending_flag'] = 'High spender'
df.loc[df['avg_product_price'] < 10, 'spending_flag'] = 'Low spender'

In [17]:
# Examine Value Count on 'spending_flag' Column
df['spending_flag'].value_counts(dropna = False)

Low spender     31770614
High spender      634245
Name: spending_flag, dtype: int64

In [18]:
# Checking Releavent Columns
df[['user_id', 'avg_product_price', 'spending_flag']].head(15)

user_id  avg_product_price spending_flag
0         1           6.367797   Low spender
1         1           6.367797   Low spender
2         1           6.367797   Low spender
3         1           6.367797   Low spender
4         1           6.367797   Low spender
5         1           6.367797   Low spender
6         1           6.367797   Low spender
7         1           6.367797   Low spender
8         1           6.367797   Low spender
9         1           6.367797   Low spender
10       15           3.980556   Low spender
11       15           3.980556   Low spender
12       15           3.980556   Low spender
13       15           3.980556   Low spender
14       15           3.980556   Low spender

# 7. Determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order”

In [19]:
# Aggregating Data with transform() Function
df['median_days_between_orders'] = df.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [20]:
# Checking Relevant Columns 
df[['user_id', 'days_since_prior_order', 'median_days_between_orders']].head(15)

user_id  days_since_prior_order  median_days_between_orders
0         1                     NaN                        20.5
1         1                    15.0                        20.5
2         1                    21.0                        20.5
3         1                    29.0                        20.5
4         1                    28.0                        20.5
5         1                    19.0                        20.5
6         1                    20.0                        20.5
7         1                    14.0                        20.5
8         1                     0.0                        20.5
9         1                    30.0                        20.5
10       15                     7.0                        10.0
11       15                     8.0                        10.0
12       15                     7.0                        10.0
13       15                     7.0                        10.0
14       15                    14.0                        10.0

In [21]:
# Creating a 'Order Frequency' Label Flag using loc Function
df.loc[df['median_days_between_orders'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'
df.loc[(df['median_days_between_orders'] <= 20) & (df['median_days_between_orders'] > 10), 'order_frequency_flag'] = 'Regular customer'
df.loc[df['median_days_between_orders'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [22]:
# Examine Value Counts on 'order_frequency_flag'
df['order_frequency_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: order_frequency_flag, dtype: int64

In [23]:
# Checking Relevant Columns of Updated Dataframe
df[['user_id', 'days_since_prior_order', 'median_days_between_orders', 'order_frequency_flag']].head(15)

user_id  days_since_prior_order  median_days_between_orders  \
0         1                     NaN                        20.5   
1         1                    15.0                        20.5   
2         1                    21.0                        20.5   
3         1                    29.0                        20.5   
4         1                    28.0                        20.5   
5         1                    19.0                        20.5   
6         1                    20.0                        20.5   
7         1                    14.0                        20.5   
8         1                     0.0                        20.5   
9         1                    30.0                        20.5   
10       15                     7.0                        10.0   
11       15                     8.0                        10.0   
12       15                     7.0                        10.0   
13       15                     7.0                        10.0   
14       15                    14.0                        10.0   

     order_frequency_flag  
0   Non-frequent customer  
1   Non-frequent customer  
2   Non-frequent customer  
3   Non-frequent customer  
4   Non-frequent customer  
5   Non-frequent customer  
6   Non-frequent customer  
7   Non-frequent customer  
8   Non-frequent customer  
9   Non-frequent customer  
10      Frequent customer  
11      Frequent customer  
12      Frequent customer  
13      Frequent customer  
14      Frequent customer

# 9. Export dataframe as a pickle file to “Prepared Data” folder

In [24]:
# Exporting the Dataframe
df.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_grouped.pkl'))